wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileManager.java

```diff
@@ -121,23 +187,85 @@ public void resolveReference(DevfileImpl devfile, FileContentProvider fileConten
     }
   }
 
-  private DevfileImpl parse(String content, ValidationFunction validationFunction)
+  private DevfileImpl parse(
+      String content, ObjectMapper mapper, Map<String, String> overrideProperties)
       throws DevfileFormatException {
-    JsonNode parsed = validationFunction.validate(content);
-
     DevfileImpl devfile;
     try {
-      devfile = objectMapper.treeToValue(parsed, DevfileImpl.class);
+      JsonNode parsed = mapper.readTree(content);
+      parsed = applyPropertiesOverride(parsed, overrideProperties);
+      schemaValidator.validate(parsed);
+      devfile = mapper.treeToValue(parsed, DevfileImpl.class);
     } catch (JsonProcessingException e) {
       throw new DevfileFormatException(e.getMessage());
+    } catch (IOException e) {
+      throw new DevfileFormatException("Unable to parse Devfile. Error: " + e.getMessage());
     }
-
     integrityValidator.validateDevfile(devfile);
     return devfile;
   }
 
-  @FunctionalInterface
-  private interface ValidationFunction {
-    JsonNode validate(String content) throws DevfileFormatException;
+  private JsonNode applyPropertiesOverride(
+      JsonNode devfileNode, Map<String, String> overrideProperties) throws DevfileFormatException {
+    for (Map.Entry<String, String> entry : overrideProperties.entrySet()) {
+      // prepare stuff
+      String[] pathSegments = entry.getKey().split("\\.");
+      if (pathSegments.length < 1) {
+        continue;
+      }
+      String lastSegment = pathSegments[pathSegments.length - 1];
```

Comment:
Similar to the discussion above, I think we should do the sanity check on the lastSegment as well. Because path like `..` would lead to an empty `pathSegment` which I don't think is valid.

Reply:
No, it can be empty only if whole path is empty which is probably not possible via url parameters